File: java/Token.java
import java.util.LinkedList;
import java.util.List;
import java.lang.String;
import java.util.regex.Pattern;

public class Token {
    // PYTHON VERSION var RE = /([^\s"]*"(?:[^"\\]*(?:\\.[^"\\]*)*)"\S*|[^\s']*'(?:[^'\\]*(?:\\.[^'\\]*)*)'\S*|\S+)/;

    // private static Pattern SPLIT_RE = Pattern.compile("([^\\s\"]*\"(?:[^\"\\\\]*(?:\\\\.[^\"\\]*)*)\"\\S*|[^\\s']*'(?:[^'\\\\]*(?:\\\\.[^'\\\\]*)*)'\\S*|\\S+)");
    private static Pattern SPLIT_RE = Pattern.compile("([^\\s\"]*\""
                                                     +"(?:[^\"\\\\]*"
                                                       +"(?:\\\\.[^\"\\\\]*)"
                                                     +"*)\"\\S*|[^\\s']*'"
                                                     +"(?:[^'\\\\]*"
                                                       +"(?:\\\\.[^'\\\\]*)"
                                                     +"*)'\\S*|\\S+)");

    enum Type {
        TOKEN_TEXT,
        TOKEN_VAR,
        TOKEN_BLOCK,
        TOKEN_COMMENT,
    };

    private Type    type;
    public String   contents;
    private int     lineno = -1;


    public Token(Type token_type, String token_contents) {
        type = token_type;
        contents = token_contents;
    }

    public String toString() {
        int maxl = contents.length();
        if (maxl > 20) 
            maxl = 20;
        return "<" + type + " token: \"" + contents.substring(0,maxl).replaceAll("\n", " ") + "\"...>";
    }

    public List<String> split_contents() {
        List<String>    split = new LinkedList<String>();

        String[]            bits = SPLIT_RE.split(contents);
        int                 idx = 0;

        while (idx < bits.length) {
            String  bit = bits[idx++];
            // Handle trnaslation-marked template pieces

            if (bit.startsWith("_(\"") || bit.startsWith("_('")) {
                    List<String>    trans_bit = new LinkedList<String>();
                    String    sentinal = bit.charAt(2) + ")";
                    
                    while (!bit.endsWith(sentinal) && idx < bits.length) {
                        bit = bits[idx++];
                        trans_bit.add(bit);
                    }
                    // Join it up...  maybe should be in the the above loop...
                    StringBuffer    sbuf = new StringBuffer();
                    for (String s : trans_bit) {
                        sbuf.append(s);
                        sbuf.append(' ');
                    }
                    bit = sbuf.toString();
            }
            split.add(bit);
        }

        return split;
    }

    public boolean empty() {
        return contents == null || contents.length() == 0; 
    }
}

File: nodejs/token.js
var sys = require('sys');

module.exports = Token;

function Token(type, contents) {
    this.type = type;
    this.contents = contents;
};
    
Token.TOKEN_TEXT    = 0;
Token.TOKEN_VAR     = 1;
Token.TOKEN_BLOCK   = 2;
Token.TOKEN_CONTENT = 3;

Token.prototype = {
    toString : function() {
        //sys.puts(this.type);
        var stype = ["Text", "Var", "Block", "Comment"];
        if (this.contents.length <= 20) {
            return '<' + stype[this.type] + ' "' + this.contents + '">';
        } else {
            return '<' + stype[this.type] + ' "' + this.contents.substr(0, 20) + '...">';
        }
    },

    split_contents : function() {
        var RE = /([^\s"]*"(?:[^"\\]*(?:\\.[^"\\]*)*)"\S*|[^\s']*'(?:[^'\\]*(?:\\.[^'\\]*)*)'\S*|\S+)/;
        var split = [];

        // sys.puts("TOKEN = '" + this.contents + "'");

        var parts = this.contents.split(RE).filter(function (s) {
            for (var i = 0; i < s.length; i++) {
                if (s[i] != ' ') {
                    return true;
                }
            }
            return false;
        });

        // sys.puts("PARTS = [" + parts + "]");

        for (var i = 0; i < parts.length; i++) {
            var bit = parts[i];
            if (bit.length == 0) continue;
            //sys.puts('DEBUG: ' + bit)
            if (bit.substr(0, 3) == '_("' || bit.substr(0,3) == "_('") {
                var sentinal = bit.charAt(2) + ')';
                var trans_bit = [bit];
                while (! bit.substr(bit.length - 3) == sentinal) {
                    bit = parts[++i];
                    trans_bit.push(bit);
                }
                bit = trans_bit.join(' ');
            }
            split.push(bit)
        }

        return split;
    }
};
